--- a/drivers/kernelsu/infra/file_wrapper.c
+++ b/drivers/kernelsu/infra/file_wrapper.c
@@ -105,7 +105,11 @@
 
 static ssize_t ksu_wrapper_iopoll(struct kiocb *kiocb, bool spin)
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0)
     return orig->f_op->iopoll(kiocb, spin);
+#else
+    return -EINVAL;
+#endif
 }
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 20, 0)
@@ -334,6 +338,7 @@
 
     loff_t res = -EINVAL;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 20, 0)
     if (remap_flags & REMAP_FILE_DEDUP) {
         res = orig->f_op->remap_file_range(file_in, pos_in, orig, pos_out, len,
                                              remap_flags);
@@ -345,6 +350,7 @@
                                              orig, pos_out, len, remap_flags);
         }
     }
+#endif
     return res;
 }
 #endif
@@ -389,7 +395,9 @@
     p->ops.read_iter = fp->f_op->read_iter ? ksu_wrapper_read_iter : NULL;
     p->ops.write_iter = fp->f_op->write_iter ? ksu_wrapper_write_iter : NULL;
     p->ops.splice_read = fp->f_op->splice_read ? ksu_wrapper_splice_read : NULL;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0)
     p->ops.iopoll = fp->f_op->iopoll ? ksu_wrapper_iopoll : NULL;
+#endif
     p->ops.fsync = fp->f_op->fsync ? ksu_wrapper_fsync : NULL;
     p->ops.fasync = fp->f_op->fasync ? ksu_wrapper_fasync : NULL;
     p->ops.lock = fp->f_op->lock ? ksu_wrapper_lock : NULL;
@@ -426,8 +434,10 @@
     p->ops.setlease = fp->f_op->setlease ? ksu_wrapper_setlease : NULL;
     p->ops.fallocate = fp->f_op->fallocate ? ksu_wrapper_fallocate : NULL;
     p->ops.show_fdinfo = fp->f_op->show_fdinfo ? ksu_wrapper_show_fdinfo : NULL;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 20, 0)
     p->ops.remap_file_range =
         fp->f_op->remap_file_range ? ksu_wrapper_remap_file_range : NULL;
+#endif
     p->ops.fadvise = fp->f_op->fadvise ? ksu_wrapper_fadvise : NULL;
 #endif
